### ISSUE TITLE
Adding in custom events 

### ISSUE THREAD

**Issue by Airick73** [gramps-project/Gramps.js#118]
I'm looking to add in custom events but I could not find anywhere to make a custom event on the Gramps.js front-end. Looking in the code for adding a new event I found it was related to `Gramps.js\src\components\GrampsjsFormSelectType.js`. In the `getTypes` function, there is a reference to `customtypes`.  I am unsure of how to add to these custom events on the front-end. 

![image](https://user-images.githubusercontent.com/72322466/194950139-b24790b4-11fd-4f12-b203-80ce6950885b.png)

I ran the gramps-web-api and successfully performed a POST for a custom event and I was able to GET the same event, so I know it is possible.

**Comment by Airick73** [comment]
I was able to find a workaround by adding additional parameters to line 51 
`return defaultTypes.concat(customTypes, 'Custom event one', 'Custom event two') //concatinating default types!!!`
I feel there is a better way to go about this.

**Comment by DavidMStraub** [comment]
Yes, it's correct that this is a feature that is available in the API but not yet implemented in the frontend.

It takes some thought what would be a good UI to implement this. Possibly a button below or next to the "select type" button for adding a new type which reveals a new text input field. If the text input field contains characters, the select type button should be greyed out.

Then on submitting the form, the new type has to be created before creating the event.

**Comment by Airick73** [comment]
The Gramps.JS architecture has been a bit intimidating to edit. I was going with an approach that would allow me to reuse as much code as possible to make this enhancement. I was thinking that I could have a text box right below the select type box that if a user were to enter text it would be serve as the type. When the add button would be clicked it would concatenate the custom event to the `types.custom` list and  create the new event simultaneously. I've provided a mock up below. 

![image](https://user-images.githubusercontent.com/72322466/195664132-d94f7c7b-3b15-4b58-aa88-0ad29b7493c1.png)

I am still having trouble understanding the code within the `render()` function specifically how the type is being stored (once selected) once the form is submitted using the add button. If I could figure that out then maybe I could use similar logic for typing in a textbox instead of using a select drop down for a `type`.

**Comment by DavidMStraub** [comment]
> I am still having trouble understanding the code within the render() function of GrampsjsFormSelectType.js , specifically how the type is being stored (once selected) once the form is submitted using the add button. If I could figure that out then maybe I could use similar logic for typing in a textbox instead of using a select drop down for a type.

Once the submit add is clicked, the `_submit` method in `GrampsViewNewEvent` is called. It posts the content of `this.data` to `/api/events/`.

What's tricky is that for the custom type one would need to have a separate `apiPost` and wait for it to resolve before posting the event. So that would require changing the `_submit` method of the parent class.

**Comment by Airick73** [comment]
>If the text input field contains characters, the select type button should be greyed out.

I was able to submit a custom event type through the textbox. It would be better as you suggested if I was able to disable the select type when one would type in the custom event text box. Also if I could require one to be mandatory if the other was disabled. So far I have not been able to produce that functionality. 

![image](https://user-images.githubusercontent.com/72322466/196211611-f972521c-0bca-4c43-bc82-5b33f978e7df.png)
![image](https://user-images.githubusercontent.com/72322466/196212599-5eaf2592-0869-495e-8521-4bee43cced6d.png)
![image](https://user-images.githubusercontent.com/72322466/196211783-b355c6d4-474a-4df7-b745-32c29ae6226f.png)

**Comment by DavidMStraub** [comment]
>  It would be better as you suggested if I was able to disable the select type when one would type in the custom event text box. Also if I could require one to be mandatory if the other was disabled. So far I have not been able to produce that functionality.

In the `@input` callback of the new text field, check the content and if non-empty, clear the select. Set a new property (`this._hasCustomType` or sth like that) to `true` and disable the select: `?disabled=${this._hasCustomType}`. Obviously, you also have to set `?required=${!this._hasCustomType}` as it's only required if the text field is empty. Then, in `checkFormValidity`, check that one of the two is non-empty, otherwise set `this.isFormValid` to `false`.

Concerning the UI: I think it would be clearer to not show the text field initially, but instead show a button that un-hides the text field. It could be labelled something like "create new custom type instead"; one could then even simply use the button event to set `this._hasCustomType`.

**Comment by DavidMStraub** [comment]
It occurred to me that actually the custom type text field should be added directly to `GrampsjsFormSelectType.js` as adding a custom type instead of selecting an existing one is not restricted to events.

**Comment by Airick73** [comment]
I actually have been adding this code to `GrampsjsFormSelectType.js` the whole time . This way it does show up throughout other areas of Gramps.js that allow you to enter a custom type as you said. Example attached below. 
![image](https://user-images.githubusercontent.com/72322466/197551631-908199fa-59cc-4cd7-87e2-0f5e955ba5b4.png)

**Comment by Airick73** [comment]
Thank you for your feedback in this issue. I've added some code for form validity by checking for both `custom-type` and `select-type` in the `isValid` function so I did not have to alter too much code.  
![image](https://user-images.githubusercontent.com/72322466/197622969-481ade55-38bc-4be6-b606-722f9c99477e.png)
Concerning the UI, I've added a button that utilizes `@click` to change `this._hasCustomEvent` to true and inserts HTML for entering a custom event as well disabling select type.
![image](https://user-images.githubusercontent.com/72322466/197623074-56f1c993-9d43-458f-b9ca-76b1534935cc.png)
![image](https://user-images.githubusercontent.com/72322466/197623128-7be977e2-5196-4871-906f-865a92d232bb.png)
I was considering making a pull request if you would like this functionality.
